New features for generation subject to constraints

TimSheard opened the pull request:
This adds additional features to the generator towards being able to generate Transactions
Adds Shrinking
Adds new kinds of constraints (demand driven to express random Tx subject to constraints)
Adds a file GenTx.hs (not yet a module) that explores what can be done with the new features.

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [ ] Any changes are noted in the `CHANGELOG.md` for affected package
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

UlfNorell committed:
Shrinking with constraints

TimSheard committed:
Added Vars, TypeReps for Tx. file genTx.hs
Added new Term 'Resrict' and Preds (:<-:), 'Member', 'NotMember', 'List', 'Choose'.
Added Count and Sizeable instances for EpochNo
Added Preds 'Maybe', and more detail to GenTx example.

JaredCorduan reviewed (approved):
:+1:

TimSheard committed:
Replaced Unique with RelLens, Add additional Solver cases.
Added PairSpec (whose meaning is a Transformer)
Added Member, NotMember, and MapMember Pred
Added module Test.Cardano.Ledger.Constrained.Scripts
Added Preds directory and stages for PParams, Universes, TxBody.
Authored by Tim Sheard and Aniket Deshpande

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -152,8 +161,19 @@ instance Show (Pred era) where
+  show (NotMember x y) = "NotMember " ++ show x ++ " " ++ show y
+  show (x :<-: y) = show x ++ " :<-: " ++ showT y
+  show (List t xs) = "List " ++ show t ++ " [" ++ showL show ", " xs ++ "]"
+  show (Choose s term xs) = unlines (("Choose " ++ show s ++ " " ++ show term) : (map showchoices xs))
```
```suggestion
  show (Choose s term xs) = unlines (("Choose " ++ show s ++ " " ++ show term) : map showchoices xs)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Ast.hs:
```diff
@@ -152,8 +161,19 @@ instance Show (Pred era) where
+    where
+      showchoices (target, ps) = showAllTarget target ++ showT target ++ " | " ++ showL show ", " ps ++ ")"
+  show (Maybe term target ps) = "Maybe " ++ show term ++ showAllTarget target ++ " | " ++ showL show ", " ps
   showList xs ans = unlines (ans : (map show xs))
```
```suggestion
  showList xs ans = unlines (ans : map show xs)
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Rewrite.hs:
```diff
@@ -159,6 +163,68 @@ removeTrivial = filter (not . trivial)
+      let (m1, subst) = fresh (m0, []) target
+          target2 = substTarget subst target
+          ps2 = map (substPred subst) preds
+       in -- (x,m2) = (Var (V (nm++"."++show m1) r No),m1+1)
```
Dead code
```suggestion
       in
```

TimSheard replied:
no longer applicable

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Shrink.hs:
```diff
@@ -0,0 +1,69 @@
+-- | Compute something satisfying the constraints that's as "close" to the original value as
+--   possible. TODO: more cleverness
+fixupVar :: Era era => Name era -> [Pred era] -> Payload era -> Maybe (Payload era)
+fixupVar v cs p = listToMaybe [p' | p' <- [p | validAssignment v p cs] ++ reverse (shrinkVar v cs p)]
```
```suggestion
fixupVar v cs p = listToMaybe $ [p | validAssignment v p cs] ++ reverse (shrinkVar v cs p)
```

lehins reviewed (approved):
:+1:

aniketd committed:
Add class Specification and first generic test specs

TimSheard committed:
Cleanup Fields (removed Access, inlined Yes)
added Terms HashD (hashData) and HashS (hashScipt) and Pair
Added type Pat and Pred ForEach
Added File Preds/TxOut.hs handles all TxOut fields in all eras
Added Universes for Data, Datum,
Replace Sums constraints with Lenses for specs (Aniket Deshpande)
Added type family witness in module Proof.hs
Added support for Conway DCert in Preds/Certs.hs
Added support for generating Transactions in Preds/Tx.hs
Added Pred for LedgerState. Renamed DPState->CertState, TxBody->Tx.
Added SubMap, Before, If, and Oneof Pred constructors
Added Pairing.hs for (Adds ExUnits) instance"
Added Terms HashD (hashData) and HashS (hashScipt) and Pairs
Added Vars, TypeReps for Tx. file genTx.hs
Added new Term 'Resrict' and Preds (:<-:), 'Member', 'NotMember', 'List', 'Choose'.
Added Count and Sizeable instances for EpochNo
Added Preds 'Maybe', and more detail to GenTx example.
Added Preds Oneof, SubMap, If, Before.
Added Stages PParams.hs, Univ.hs, Certs.hs, CertState.hs LedgerState.hs, Tx.hs
Updated for Eras Alonzo, Mary, Shelley
Updated to compile with GHC 9.6.2

lehins commented on eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs:
```diff
@@ -241,7 +241,7 @@ validateCollateralContainsNonADA txBody utxoCollateral =
       case txBody ^. collateralReturnTxBodyL of
         SJust retTxOut
-          | not (Val.isAdaOnly colbal) ->
+          | {- not -} (Val.isAdaOnly colbal) ->
```
Why is this negated????

TimSheard replied:
Now I am not so sure.

TimSheard replied:
So I  put the not back in

lehins commented on eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Deleg.hs:
```diff
@@ -342,7 +343,17 @@ delegationTransition = do
+                            ++ unlines (map show (Map.toList cm))
+                        )
+                        $ MIRProducesNegativeUpdate
+                     )
```
Left over debug tracing:
```suggestion
all (>= mempty) cm ?! MIRProducesNegativeUpdate
```

TimSheard replied:
done

lehins commented on libs/cardano-ledger-test/cardano-ledger-test.cabal:
```diff
@@ -74,12 +86,15 @@ library
         cardano-ledger-api,
         cardano-ledger-babbage:{cardano-ledger-babbage, testlib},
         cardano-ledger-babbage-test,
+        cardano-ledger-byron,
```
Why do we need this dependency on `cardano-ledger-byron` and on `cardano-crypto*` packages above?

TimSheard replied:
Because we are stuck with Byron Adresses forever, I felt I needed to add tests that had Byron addresses. To generate Byron Verification and Singing key pairs, I need these two modues.
import qualified Cardano.Chain.Common as Byron
import qualified Cardano.Crypto.Signing as Byron
Both if there are in  'cardano-ledger-byron' so that is why   cardano-ledger-byron is in the 'build depends' of cardano-ledger-test.

aniketd reviewed (approved):
LGTM! :raised_hands:

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/TypeRep.hs:
```diff
@@ -440,19 +781,120 @@ genSizedRep _ RewardR = arbitrary
+gen32ByteString :: Gen ByteString
+gen32ByteString = do
+  list <- vectorOf 32 (elements ["1", "2", "3", "4", "5", "6", "7", "8", "9", "0"])
+  pure (List.foldl' (<>) mempty list)
```
Here is a generator you can use for generating random bytestrings: https://github.com/input-output-hk/cardano-ledger/blob/a4895859484ed34498c80c1d737dc2c4501f6fee/libs/cardano-ledger-binary/testlib/Test/Cardano/Ledger/Binary/Arbitrary.hs#L229-L230
```suggestion
gen32ByteString = genByteString 32
```

TimSheard replied:
Thank you for pointing this out.

lehins commented on libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Proof.hs:
```diff
@@ -363,3 +412,96 @@ instance Shaped Proof any where
+
+data TxCertWit era where
+  TxCertShelleyToBabbage :: (TxCert era ~ ShelleyTxCert era, ShelleyEraTxCert era, ProtVerAtMost era 8) => TxCertWit era
+  TxCertConwayToConway :: (TxCert era ~ ConwayTxCert era, ConwayEraTxCert era, ProtVerAtLeast era 9) => TxCertWit era
```
I am pretty sure you don't need this constraint
```suggestion
  TxCertConwayToConway :: (TxCert era ~ ConwayTxCert era, ConwayEraTxCert era) => TxCertWit era
```

TimSheard replied:
done

lehins reviewed (approved):
It'd take me a week to review this PR. 
Skimmed through it. Looks good at a high level.
